src/univariate/discrete/bernoulli.jl

```diff
@@ -113,44 +113,41 @@ rand(rng::AbstractRNG, d::Bernoulli) = rand(rng) <= succprob(d)
 
 #### MLE fitting
 
-struct BernoulliStats <: SufficientStats
-    cnt0::Float64
-    cnt1::Float64
-
-    BernoulliStats(c0::Real, c1::Real) = new(Float64(c0), Float64(c1))
+struct BernoulliStats{C<:Real} <: SufficientStats
+    cnt0::C
+    cnt1::C
 end
 
+BernoulliStats(c0::Real, c1::Real) = BernoulliStats(promote(c0, c1)...)
+
 fit_mle(::Type{<:Bernoulli}, ss::BernoulliStats) = Bernoulli(ss.cnt1 / (ss.cnt0 + ss.cnt1))
 
 function suffstats(::Type{<:Bernoulli}, x::AbstractArray{T}) where T<:Integer
-    n = length(x)
     c0 = c1 = 0
-    for i = 1:n
+    for i = eachindex(x)
         @inbounds xi = x[i]
         if xi == 0
             c0 += 1
         elseif xi == 1
             c1 += 1
         else
-            throw(DomainError())
+            throw(DomainError(xi, "the sampled value must be 0 or 1"))
         end
     end
     BernoulliStats(c0, c1)
 end
 
-function suffstats(::Type{<:Bernoulli}, x::AbstractArray{T}, w::AbstractArray{Float64}) where T<:Integer
-    n = length(x)
-    length(w) == n || throw(DimensionMismatch("Inconsistent argument dimensions."))
-    c0 = c1 = 0
-    for i = 1:n
+function suffstats(::Type{<:Bernoulli}, x::AbstractArray{T}, w::AbstractArray{<:Real}) where T<:Integer
+    c0 = c1 = zero(eltype(w))
+    for i = eachindex(x, w)
         @inbounds xi = x[i]
         @inbounds wi = w[i]
```

Comment:
This would be breaking since currently we don't enforce that `x` and `w` have the same indices (nor that they have the same shape). Instead we should use something like
```suggestion
    length(w) == n || throw(DimensionMismatch("inconsistent argument dimensions"))
    c0 = c1 = zero(eltype(w))
    for xi, wi in zip(x, w)
```

Reply:
> Would eachindex(vec(x), vec(w)) make it better then?

IMO that's less clear than just iterating over the two arrays.

>  The fact that x and w can have different shapes seems to me a possible source of bugs though.

We might want to restrict it at some point but it's a breaking change and hence should be postponed at least until we prepare the next breaking release.

Reply:
Could you give me an example of where `eachindex` fails whereas the old implementation worked?
For instance, `eachindex([1,2,3,4], [1 2; 3 4])` works fine and returns `Base.OneTo(4)`. I can't see an example that breaks.